File: boosting/src/data/dataholder/ObjectDataHolder.java
package data.dataholder;

import data.Data;
import data.DataReader;

public class ObjectDataHolder<T extends Data> extends DataHolder<T> {

    public ObjectDataHolder(DataReader<T> dataReader, double trainPercent, boolean normalize, boolean shuffle) {
        super(dataReader, trainPercent, normalize, shuffle);
    }

    @Override
    protected void normalize() {
        if (data == null || data.isEmpty()) {
            return;
        }
        int vectorSize = data.get(0).asVector().length;
        for (int i = 0; i < vectorSize; i++) {
            double min = Double.POSITIVE_INFINITY, max = Double.NEGATIVE_INFINITY;
            for (Data data : this.data) {
                min = Math.min(min, data.asVector()[i]);
                max = Math.max(max, data.asVector()[i]);
            }
            for (Data data : this.data) {
                double value = data.asVector()[i];
                data.asVector()[i] = (value - min) / (max - min);
            }
        }
    }

    @Override
    public int getVectorSize() {
        if (data == null || data.isEmpty()) {
            return 0;
        }
        return data.get(0).asVector().length;
    }
}

File: boosting/src/data/dataholder/ObjectDataHolder.kt
package data.dataholder

import data.Data
import data.DataReader

class ObjectDataHolder<T : Data>(dataReader: DataReader<T>, trainPercent: Double, normalize: Boolean, shuffle: Boolean) : DataHolder<T>(dataReader, trainPercent, normalize, shuffle) {

    override fun normalize() {
        for (i in 0..vectorSize - 1) {
            var min = Double.POSITIVE_INFINITY
            var max = Double.NEGATIVE_INFINITY
            for (data in this.data) {
                min = Math.min(min, data.vector[i])
                max = Math.max(max, data.vector[i])
            }
            for (data in this.data) {
                val value = data.vector[i]
                data.vector[i] = (value - min) / (max - min)
            }
        }
    }

    override val vectorSize: Int
        get() {
            return data[0].vector.size
        }
}
